finetune/base.py

```diff
@@ -92,34 +91,73 @@ def cleanup():
         if self.config.debugging_logs:
             os.environ["TF_CPP_MIN_LOG_LEVEL"] = "0"
             tf_logging.set_verbosity(tf_logging.DEBUG)
+        self.check_gpu_for_fp16()
 
-    def resolve_config(self, **kwargs):
-        assert_valid_config(**kwargs)
+    def check_gpu_for_fp16(self):
+        if not gpu_info(self._get_estimator_config().session_config)["fp16_inference"]:
+            if self.config.float_16_predict or self.config.mixed_precision:
+                LOGGER.warning(
+                    "This GPU does not support float 16 ops but they were requested in the config. They are being turned off."
+                )
+                self.config = self.resolve_config(no_fp16=True)
+
+    def resolve_config(self, no_fp16=False, **kwargs):
+        self.config_overrides.update(kwargs)
+        assert_valid_config(**self.config_overrides)
         config = get_default_config()
         config.base_model = kwargs.get("base_model", config.base_model)
 
         auto_keys = []
-        config.update(kwargs)
+        config.update(self.config_overrides)
         for k, v in config.items():
             if isinstance(v, str) and v.lower() == "auto":
                 auto_keys.append(k)
-            elif k in kwargs:
+            elif k in self.config_overrides:
                 continue
             elif k in config.base_model.settings:
                 config[k] = config.base_model.settings[k]
 
-        overrides = config.base_model.get_optimal_params(config)
-
-        if not issubclass(config.base_model, _BaseBert) and config.float_16_predict:
-            LOGGER.warning("float_16_predict only supported by bert based models")
+        # This has to be here before optimal_params are derived because some are dependant on these values.
+        if (not issubclass(config.base_model, _BaseBert) or no_fp16) and (
+            config.float_16_predict == True or config.mixed_precision == True
+        ):
+            LOGGER.warning(
+                "float_16_predict and mixed_precision only supported by bert based models and Volta or newer GPUs"
+            )
             config.float_16_predict = False
+            config.mixed_precision = False
+
+        if (
+            not issubclass(config.base_model, _BaseBert) or no_fp16
+        ) and "fp16" in config.optimize_for:
```

Comment:
Interesting -- what's the rationale behind making fp16 be part of the optimize_for string?

Reply:
Primarily just so we don't have to add a another tier of config defaults.

Reply:
Resolution is already quite complex. I would agree with potentially renaming this from optimize_for to something like "preset_config" or something but I don't think we should change these automatically when you set mixed_precision = True or anything.